PokemonAPI/PokemonAPI.BusinessLayer/Implementations/DomainServices/BattleService.cs

```diff
@@ -0,0 +1,111 @@
+﻿using Microsoft.EntityFrameworkCore;
+using PokemonAPI.BusinessLayer.Interfaces;
+using PokemonAPI.BusinessLayer.Models;
+using PokemonAPI.DomainLayer.Entities;
+using PokemonAPI.PersistenceLayer.Interfaces;
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace PokemonAPI.BusinessLayer.Implementations.DomainServices
+{
+    internal class BattleService : IBattleService
+    {
+
+        private readonly IRepository<Pokemon> _repository;
+
+        public BattleService(IRepository<Pokemon> repository)
+        {
+            _repository = repository;
+        }
+
+        public async Task<BattleResult> Battle(Guid firstPokemonId, Guid secondPokemonId, CancellationToken cancellationToken = default)
+        {
+            var  firstPokemon = await _repository.Get(
+                filter: dbPokemon => dbPokemon.Id == firstPokemonId,
+                 include: source => source
+                    .Include(x => x.Abilities)
+                    .ThenInclude(x => x.Ability),
+                cancellationToken: cancellationToken
+            );
+
+            var secondPokemon = await _repository.Get(
+                filter: dbPokemon => dbPokemon.Id == secondPokemonId,
+                 include: source => source
+                    .Include(x => x.Abilities)
+                    .ThenInclude(x => x.Ability),
+                cancellationToken: cancellationToken
+            );
+
+            int GoesFirst(Pokemon firstPokemon, Pokemon secondPokemon) {
+                if (firstPokemon.BaseStats.Speed > secondPokemon.BaseStats.Speed)
+                {
+                    return 1;
+                }
+                else if (firstPokemon.BaseStats.Speed < secondPokemon.BaseStats.Speed)
+                {
+                    return 2;
+                }
+                else
+                {
+                    return 1;
+                }
+            }
+
+            int first = GoesFirst(firstPokemon, secondPokemon);
+            int firstPokeHealth = firstPokemon.BaseStats.HealthPoints;
+            int secondPokeHealth = secondPokemon.BaseStats.HealthPoints;
+            bool specialFirst = false;
+            bool specialSecond = false;
+
+            do
+            {
+                if (first == 1)
+                {
+                    firstPokeHealth = firstPokemon.BaseStats.SpecialAttack - secondPokemon.BaseStats.SpecialDefense;
+                    first = 2;
+                    specialFirst = true;
+                }
+                else
+                {
+                    secondPokeHealth = secondPokemon.BaseStats.SpecialDefense - firstPokemon.BaseStats.SpecialAttack;
+                    first = 1;
+                    specialSecond = true;
+                }
+            }
+            while (specialFirst && specialSecond && firstPokeHealth > 0 || secondPokeHealth > 0);
```

Comment:
We can do this without Do while loop

Suppose this: Hp1 + SDef1 - SAtt2  = remainingHealth1
And vice versa

Also here on line 65 and 71 you are overwriting your poke health

So say your HP1 = 100, SDef1 = 20 and SAttack2 = 20 
In this loop you will do following... HP1 = SDef1 - SAttack2 = 0

You will completely remove HP1 = 100.., and turn it into 0. But that is not true... the first pokemon still has 100 HP left. He used his special defense that is equal to the special attack of the second pokemon.

I may be wrong since I'm tired. Please do correct me.